transports/janus_pfunix.c

```diff
@@ -0,0 +1,526 @@
+	/* FIXME Should we really close the connection in case of a timeout? */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) != NULL && client->fd > -1) {
+		/* Shutdown the client socket */
+		shutdown(client->fd, SHUT_WR);
+	}
+	janus_mutex_unlock(&clients_mutex);
+}
+
+
+/* Thread */
+void *janus_pfunix_thread(void *data) {
+	JANUS_LOG(LOG_INFO, "Unix Sockets thread started\n");
+
+	int fds = 0;
+	struct pollfd poll_fds[1024];	/* FIXME Should we allow for more clients? */
+	char buffer[BUFFER_SIZE];
+
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		/* Prepare poll list of file descriptors */
+		fds = 0;
+		/* Writeable monitor */
+		poll_fds[fds].fd = write_fd[0];
+		poll_fds[fds].events = POLLIN;
+		fds++;
+		if(pfd > -1) {
+			/* Janus API */
+			poll_fds[fds].fd = pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		if(admin_pfd > -1) {
+			/* Admin API */
+			poll_fds[fds].fd = admin_pfd;
+			poll_fds[fds].events = POLLIN;
+			fds++;
+		}
+		/* Iterate on available clients, to see if we need to POLLIN or POLLOUT too */
+		janus_mutex_lock(&clients_mutex);
+		GHashTableIter iter;
+		gpointer value;
+		g_hash_table_iter_init(&iter, clients_by_fd);
+		while(g_hash_table_iter_next(&iter, NULL, &value)) {
+			janus_pfunix_client *client = value;
+			if(client->fd > -1) {
+				poll_fds[fds].fd = client->fd;
+				poll_fds[fds].events = g_async_queue_length(client->messages) > 0 ? POLLIN | POLLOUT : POLLIN;
+				fds++;
+			}
+		}
+		janus_mutex_unlock(&clients_mutex);
+
+		/* Start polling */
+		int res = poll(poll_fds, fds, -1);
+		if(res == 0)
+			continue;
+		if(res < 0) {
+			JANUS_LOG(LOG_ERR, "poll() failed\n");
+			break;
+		}
+		for(int i=0; i<fds; i++) {
+			if((poll_fds[i].revents & POLLOUT) == POLLOUT) {
+				/* Find the client from its file descriptor */
+				janus_mutex_lock(&clients_mutex);
+				janus_pfunix_client *client = g_hash_table_lookup(clients_by_fd, GINT_TO_POINTER(poll_fds[i].fd));
+				if(client != NULL) {
+					char *payload = NULL;
+					while((payload = g_async_queue_try_pop(client->messages)) != NULL) {
+						int res = 0;
+						do {
+							res = write(client->fd, payload, strlen(payload));
+						} while(res == -1 && errno == EINTR);
+						/* FIXME Should we check if sent everything? */
+						JANUS_LOG(LOG_HUGE, "Written %d/%zu bytes on %d\n", res, strlen(payload), client->fd);
+						g_free(payload);
+					}
+				}
+				janus_mutex_unlock(&clients_mutex);
+			}
+			if((poll_fds[i].revents & POLLIN) == POLLIN) {
+				if(poll_fds[i].fd == write_fd[0]) {
+					/* Read and ignore: we use this to unlock the poll if there's data to write */
+					res = read(poll_fds[i].fd, buffer, BUFFER_SIZE);
+				} else if(poll_fds[i].fd == pfd || poll_fds[i].fd == admin_pfd) {
+					/* Janus/Admin API: accept the new client */
+					struct sockaddr_un address;
+					socklen_t address_length = 0;
```

Comment:
Fair point, I recalled this was actually only filled by the system call, but it indeed needs to be initialized to the size of the struct itself first. Will fix.